scripts/add_rtp_task_resource_records.py

```diff
@@ -157,11 +240,83 @@ def query_slurm_db(jobid):
+                    )
+                    if tpe_query.count() == 0:
+                        # Some JDs are missing from RTPTaskMultipleProcessEvent
+                        # If a JD is missing, use the last attempt
+                        task = mult_attempts.order_by(RTPTaskMultipleJobID.start_time)[
+                            -1
+                        ]
+                    else:
+                        tpe_tasks = tpe_query.filter(
+                            (
+                                RTPTaskMultipleProcessEvent.obsid_start
+                                == task.obsid_start
+                            )
+                            & (RTPTaskMultipleProcessEvent.task_name == task.task_name)
+                        )
+                        finished_tasks = tpe_tasks.filter(
+                            RTPTaskMultipleProcessEvent.event == "finished"
+                        )
+                        if finished_tasks.count() > 0:
+                            # Only add rows for finished tasks
+                            finish_time = finished_tasks[-1].time
+                            started_tasks = tpe_tasks.filter(
+                                RTPTaskMultipleProcessEvent.event == "started"
+                            )
+                            start_times = started_tasks.with_entities(
+                                RTPTaskMultipleProcessEvent.time
+                            )
+                            start_times = start_times.order_by(
+                                RTPTaskMultipleProcessEvent.time
+                            )
+                            start_times = np.array(
+                                [t[0] for t in start_times.all()], dtype=int
+                            )
+                            start_time_ind = np.argmin(
+                                np.abs(finish_time - start_times)
+                            )
+                            start_time_tpe = start_times[start_time_ind]
+                            mult_times = mult_attempts.with_entities(
+                                RTPTaskMultipleJobID.start_time
+                            )
+                            mult_times = mult_times.order_by(
+                                RTPTaskMultipleJobID.start_time
+                            )
+                            mult_times = np.array(
+                                [t[0] for t in mult_times.all()], dtype=int
+                            )
+                            closest_time_ind = np.argmin(
+                                np.abs(start_time_tpe - mult_times)
+                            )
+                            closest_time = int(mult_times[closest_time_ind])
+                            task = mult_attempts.filter(
+                                RTPTaskMultipleJobID.start_time == closest_time
+                            )
+                            task = task[0]
```

Comment:
It looks like this logic is very similar to above (lines 192-222). I wonder if it's possible to harmonize them into a single function, where the table type is passed in? It looks like the queries columns (`event`, `time`, and `start_time`) are common between the two tables.

Reply:
I think it might be possible to put this logic into a function.  I'll take a look at that over the next few days.

Reply:
There's an example of a function that runs similar queries on different tables in `mc_session._time_filter` in case it's useful.